jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,285 @@
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - action_mask: jax array (bool) of shape (4,), indicating which actions are valid
+            in the current state of the environment.
+
+    - action: int32, representing the direction to move the empty tile
+        (up, down, left, right)
+
+    - reward: float, a dense reward is provided based on the arrangement of the tiles.
+        It equals the negative sum of the boolean difference between
+        the current state of the puzzle and the goal state (correctly arranged tiles).
+        Each incorrectly placed tile contributes -1 to the reward.
+
+    - episode termination: if the puzzle is solved.
+
+    - state: `State`
+        - puzzle: jax array (int32) of shape (N, N), representing the current state of the puzzle.
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - key: jax array (uint32) of shape (2,), random key used to generate random numbers
+            at each step and for auto-reset.
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        time_limit: int = 500,
+        viewer: Optional[Viewer[State]] = None,
+    ) -> None:
+        """Instantiate a `SlidingTilePuzzle` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` which generates puzzles with
+                a size of 5x5.
+            reward_fn: RewardFn whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action and the next state.
+                Implemented options are [`DenseRewardFn`, `SparseRewardFn`].
+                Defaults to `DenseRewardFn`.
+            time_limit: maximum number of steps before the episode is terminated.
+            viewer: environment viewer for rendering.
+        """
+        self.generator = generator or RandomGenerator(grid_size=5)
+        self.reward_fn = reward_fn or DenseRewardFn()
+
+        self.time_limit = time_limit
+
+        # Create viewer used for rendering
+        self._env_viewer = viewer or SlidingTilePuzzleViewer(name="SlidingTilePuzzle")
+        self.solved_puzzle = jnp.arange(self.generator.grid_size**2).reshape(
+            (self.generator.grid_size, self.generator.grid_size)
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+            step_count=0,
+        )
```

Comment:
The convention in jumanji is that the generator returns the State object. Is is possible to have the generator return 
```python
State(
            puzzle=puzzle,
            empty_tile_position=empty_tile_position,
            key=key,
            step_count=0,
        )
```
directly?